iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/datastructure/TVList.java

```diff
@@ -252,27 +406,22 @@ protected void releaseLastTimeArray() {
   }
 
   public int delete(long lowerBound, long upperBound) {
-    int newSize = 0;
-    maxTime = Long.MIN_VALUE;
+    int deletedNumber = 0;
+    long maxTime = Long.MIN_VALUE;
+    long minTime = Long.MAX_VALUE;
     for (int i = 0; i < rowCount; i++) {
       long time = getTime(i);
-      if (time < lowerBound || time > upperBound) {
-        set(i, newSize++);
+      if (time >= lowerBound && time <= upperBound) {
+        int originRowIndex = getValueIndex(i);
+        int arrayIndex = originRowIndex / ARRAY_SIZE;
+        int elementIndex = originRowIndex % ARRAY_SIZE;
+        markNullValue(arrayIndex, elementIndex);
+        deletedNumber++;
```

Comment:
May check if the time is already deleted.

Reply:
ok

Reply:
Done